<img src="./doc/design/nexgendx_logo.png" width=87%>

```py
[bash]~$ python3
>>> import nexgendx as ngs
>>> # simple command prompt example
>>> tumor_genome = ngs.teraget(, con=ngs.sqlalcmy(...))
>>> # do more
>>> 
>>> very 
>>> basic
>>> nice illustrative
>>> example 
>>> load / calc stats / visualize 
>>> :Dctrl+d
ctrl+c
```

<h1 align="center" underline=0>
Next-gen sequencing analysis tools for<br/>
circulating cell-free DNA.
</h1>

<br>

<br>

1. this
2. that
3. these
 - etc
 - etc 

<br>

## Getting Started

Dependencies:
 - <a href="">``bedtools``</a>
 - <a href="">``BWA``</a>
 
> from <a href="">htslib</a>: 
 - ``faidx`` + ``samtools`` + ``bcftools``
 
<br>

> from <a href="">the Broad Institute</a>: 
 - ``Picard`` + ``GATK(-Protected)`` java suite
 
<br>

> from <a href="">the Python library collection</a>:
> (``pip`` or ``conda`` recommended for install; *=optional)
 - ``asyncio`` + ``asycnpg``* + ``concurrent.futures``*
 - ``blaze``* 
 - ``dask``*
 - ``django``* + ``flask``* + ``bottle``* + ``sanic``*
 - ``ggplot``
 - ``matplotlib2.0``
 - ``numba``*
 - ``numpy`` + ``scipy``
 - ``pandas``
 - ``plotly``
 - ``seaborn``
 - ``SQLAlchemy``*
<br>

Example datasets:
> <a href="">FTP download / AWS access from NCBI</a>
>  * Genome in a Bottle Consortium pilot genome reference standard NA12878
>  * <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4896128/">12X-NGS-platform (BioNano Genomics, Complete Genomics paired-end and LFR, Ion Proton exome, Oxford Nanopore, Pacific Biosciences, SOLiD, 10X Genomics GemCode WGS, and Illumina exome and WGS paired-end, mate-pair, and synthetic long reads)</a> highly orthogonally validated genotype determination for high-confidence regions (2016), useful as a new "gold standard" for benchmarking NGS error (i.e. false-positives rate, sensitivity/specificity, etc.) in an unbiased fashion.
>  * See Jupyter notebooks folder under ``./examples`` for some basic analysis workflow demonstrations that are easily replicated with NA12878, harnessing the ``nexgendx`` tool suite for data I/O, stats analysis, custom pipeline setup, launch & QC, and visualization. 


## License 

NOT INTENDED FOR COMMERCIAL USE !

<big>®</big> All rights reserved. Free to copy, modify, use for personal analysis/education, share with friends, 
but not patent, sell, distribute on large-scale, use for any commerical/for-profit purposes, etc. 
(see <a href="">my LICENSE</a>, <a href="">GATK LICENSE</a>, & <a href="">more</a>)

<div align="center">

<br>

``@+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++``
``𝝘𝝝𝝝𝝘𝝩𝝝𝝘𝝖𝝝𝝩𝝩𝝝𝝖𝝩𝝖𝝖𝝝𝝩𝝩𝝝𝝩𝝩𝝩𝝝𝝘𝝖𝝝𝝩𝝩𝝝𝝩𝝖𝝖𝝝𝝩𝝩𝝝𝝩𝝖𝝖𝝝𝝖𝝖𝝝𝝩𝝩𝝝𝝩𝝖𝝖𝝝𝝩𝝩𝝝𝝩𝝩𝝩𝝝𝝘𝝩𝝝𝝘𝝖𝝘𝝖𝝖𝝩𝝝𝝝𝝘``
``++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++``
``DAEEEEDGGGIIIEJKKKJLKPKGGGGGGGGLGLGLKHHHHHKGGDJHHHHKGGDEJKKKJLKPKGGGGGKAAAAAABBJ``
``@+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++``

<br>

---------

<img src="./doc/design/nucleosome_spacewaves_1.gif">

</div>

## Tools for the nucleosome space.

<br>

<div align="center">

``⊰♘𝚿⚔⊱ 2017 | J.Collins | ⚕ nexgendx.``      ➉❶①⓪🄵🅡ⓔⓔ🄲🅾ⓓⓔ㊝①➓➀

~~~ py
(1*2^8 + 0*2^7 + 1*2^6 + 1*2^5 + 0*2^4 + 1*2^3 + 1*2^2 + 0*2^1 + 1*2^0) = 365
~~~

<h6>[logo image source]:</h6>
<h4>See <a href="https://pdb101.rcsb.org/motm/7">Protein DataBank's Educational Portal</a> for more about nucleosomes.</h4>

<br>

## 𐃠

</div>






